applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp

```diff
@@ -0,0 +1,743 @@
+        }
+
+        // initialize the modal matrix
+        auto& pModalMatrix = this->pGetModalMatrix();
+        auto& rModalMatrix = *pModalMatrix;
+        const std::size_t n_modes = rProcessInfo[EIGENVALUE_VECTOR].size();
+        if( rModalMatrix.size1() != system_size || rModalMatrix.size2() != n_modes )
+            rModalMatrix.resize( system_size, n_modes, false );
+        rModalMatrix = ZeroMatrix( system_size, n_modes );
+
+        boost::timer modal_matrix_build_time;
+        for( std::size_t i = 0; i < n_modes; ++i )
+        {
+            for( ModelPart::NodeIterator itNode = rModelPart.NodesBegin(); itNode != rModelPart.NodesEnd(); itNode++ )
+            {
+                ModelPart::NodeType::DofsContainerType& node_dofs = itNode->GetDofs();
+                const std::size_t n_node_dofs = node_dofs.size();
+                Matrix& rNodeEigenvectors = itNode->GetValue(EIGENVECTOR_MATRIX);
+
+                if( node_dofs.IsSorted() == false )
+                {
+                    node_dofs.Sort();
+                }
+
+                for( std::size_t j = 0; j < n_node_dofs; ++j )
+                {
+                    auto itDof = std::begin(node_dofs) + j;
+                    rModalMatrix(itDof->EquationId(), i) = rNodeEigenvectors(i, j);
+                }
+            }
+        }
+
+        if (BaseType::GetEchoLevel() > 0 && rank == 0)
+        {
+            std::cout << "modal_matrix_build_time : " << modal_matrix_build_time.elapsed() << std::endl;
+        }
+
+        // get the damping coefficients if they exist
+        for( ModelPart::PropertiesIterator itProperty = rModelPart.PropertiesBegin(); itProperty != rModelPart.PropertiesEnd(); itProperty++ )
+        {
+            if( itProperty->Has(SYSTEM_DAMPING_RATIO) )
+            {
+                mSystemDamping = itProperty->GetValue(SYSTEM_DAMPING_RATIO);
+            }
+            
+            if( itProperty->Has(RAYLEIGH_ALPHA) && itProperty->Has(RAYLEIGH_BETA) )
+            {
+                mRayleighAlpha = itProperty->GetValue(RAYLEIGH_ALPHA);
+                mRayleighBeta = itProperty->GetValue(RAYLEIGH_BETA);
+            }
+        }
+
+        // compute the effective material damping if required
+        if( mUseMaterialDamping )
+        {
+            // throw an error, if no submodelparts are present
+            if( rModelPart.NumberOfSubModelParts() < 1 )
+                KRATOS_ERROR << "No submodelparts detected!" << std::endl;
+            
+            //initialize all required variables
+            rModelPart.GetProcessInfo()[BUILD_LEVEL] = 2;
+            SparseMatrixType* AuxStiffnessMatrix;
+            SparseMatrixPointerType temp_stiffness_matrix;
+            mMaterialDampingRatios = ZeroVector( n_modes );
+            
+            //initialize dummy vectors
+            auto pDx = SparseSpaceType::CreateEmptyVectorPointer();
```

Comment:
afaik this is the correct way of doing what you do at the beginning of this file
right @RiccardoRossi ?

Reply:
correct